bin/livy-server

```diff
@@ -18,37 +18,160 @@
 
-set -e
 
-export LIVY_HOME=$(cd $(dirname $0)/.. && pwd)
-cd $LIVY_HOME
+usage="Usage: livy-server (start|stop)"
 
-LIBDIR="$LIVY_HOME/jars"
-if [ ! -d "$LIBDIR" ]; then
-  LIBDIR="$LIVY_HOME/server/target/jars"
-fi
-if [ ! -d "$LIBDIR" ]; then
-  echo "Could not find Livy jars directory." 1>&2
+# Find the java binary
+if [ -n "${JAVA_HOME}" ]; then
+  RUNNER="${JAVA_HOME}/bin/java"
+elif [ `command -v java` ]; then
+  RUNNER="java"
+else
+  echo "JAVA_HOME is not set" >&2
   exit 1
 fi
 
+export LIVY_HOME=$(cd $(dirname $0)/.. && pwd)
 if [ -z "$LIVY_CONF_DIR" ]; then
   LIVY_CONF_DIR="$LIVY_HOME/conf"
 fi
 
-LIVY_CLASSPATH="$LIBDIR/*:$LIVY_CONF_DIR"
-
-if [ -n "$SPARK_CONF_DIR" ]; then
-  LIVY_CLASSPATH="$LIVY_CLASSPATH:$SPARK_CONF_DIR"
+if [ -f "${LIVY_CONF_DIR}/livy-env.sh" ]; then
+  # Promote all variable declarations to environment (exported) variables
+  set -a
+  . "${LIVY_CONF_DIR}/livy-env.sh"
+  set +a
 fi
-if [ -n "$HADOOP_CONF_DIR" ]; then
-  LIVY_CLASSPATH="$LIVY_CLASSPATH:$HADOOP_CONF_DIR"
+
+if [ -z "$LIVY_IDENT_STRING" ]; then
+  export LIVY_IDENT_STRING="$USER"
 fi
-if [ -n "$YARN_CONF_DIR" ]; then
-  LIVY_CLASSPATH="$LIVY_CLASSPATH:$YARN_CONF_DIR"
+
+if [ -z "$LIVY_PID_DIR" ]; then
+  LIVY_PID_DIR=/tmp
 fi
+pid="$LIVY_PID_DIR/livy-$LIVY_IDENT_STRING-server.pid"
+
+livy_rotate_log () {
+  log=$1
+  num=${2:-5}
+  if [ -f "$log" ]; then # rotate logs
+	while [ $num -gt 1 ]; do
+	  prev=`expr $num - 1`
+	  [ -f "$log.$prev" ] && mv "$log.$prev" "$log.$num"
+	  num=$prev
+	done
+    mv "$log" "$log.$num"
+  fi
+}
+
+create_dir() {
+  dir_name=$1
+  dir_variable=$2
+  mkdir -p $dir_name
+  touch $dir_name/.test > /dev/null 2>&1
+  TEST_DIR=$?
+  if [ "${TEST_DIR}" = "0" ]; then
+    rm -f "$dir_name"/.test
+  else
+    echo "$USER doesn't have permission to write to directory: $dir_name which is $dir_variable"
+    exit 1
+  fi
+}
+
+start_livy_server() {
+
+  LIBDIR="$LIVY_HOME/jars"
+  if [ ! -d "$LIBDIR" ]; then
+    LIBDIR="$LIVY_HOME/server/target/jars"
+  fi
+  if [ ! -d "$LIBDIR" ]; then
+    echo "Could not find Livy jars directory." 1>&2
+    exit 1
+  fi
+
+  LIVY_CLASSPATH="$LIBDIR/*:$LIVY_CONF_DIR"
+
+  if [ -n "$SPARK_CONF_DIR" ]; then
+    LIVY_CLASSPATH="$LIVY_CLASSPATH:$SPARK_CONF_DIR"
+  fi
+  if [ -n "$HADOOP_CONF_DIR" ]; then
+    LIVY_CLASSPATH="$LIVY_CLASSPATH:$HADOOP_CONF_DIR"
+  fi
+  if [ -n "$YARN_CONF_DIR" ]; then
+    LIVY_CLASSPATH="$LIVY_CLASSPATH:$YARN_CONF_DIR"
+  fi
+
+  command="$RUNNER $LIVY_SERVER_JAVA_OPTS -cp $LIVY_CLASSPATH:$CLASSPATH com.cloudera.livy.server.LivyServer"
+
+  if ! [ -z "$old_livy_launch" ]; then
+    exec $command
+  else
+    # get log directory
+    if [ -z "$LIVY_LOG_DIR" ]; then
+      export LIVY_LOG_DIR="${LIVY_HOME}/logs"
+    fi
+    create_dir $LIVY_LOG_DIR "LIVY_LOG_DIR"
+    create_dir $LIVY_PID_DIR "LIVY_PID_DIR"
+    log="$LIVY_LOG_DIR/livy-$LIVY_IDENT_STRING-server.out"
+    # Set default scheduling priority
+    if [ -z "$LIVY_NICENESS" ]; then
```

Comment:
nit: this sort of code can be rewritten as `FOO=${FOO:-DEFAULT_VALUE}` in bash. Also this particular variable doesn't need to be exported.

(Same comments apply to `LIVY_LOG_DIR` and other places.)